File: app/src/main/java/app/fynnjason/fivesingletondemo/java/InnerStaticSingleton.java
package app.fynnjason.fivesingletondemo.java;

/**
 * Created by FynnJason on 2017/12/25.
 * 静态内部类方式，比较推荐的写法
 */

public class InnerStaticSingleton {
    private InnerStaticSingleton() {

    }

    private static class Holder {
        private static InnerStaticSingleton INSTANCE = new InnerStaticSingleton();
    }

    public static InnerStaticSingleton getInstance() {
        return Holder.INSTANCE;
    }
}

File: app/src/main/java/app/fynnjason/fivesingletondemo/kotlin/InnerStaticSingleton.kt
package app.fynnjason.fivesingletondemo.kotlin

/**
 * Created by FynnJason on 2017/12/25.
 * 静态内部类方式，比较推荐的写法
 */
class InnerStaticSingleton private constructor() {
    companion object {
        fun getInstance() = Holder.INSTANCE

    }

    private object Holder {
        val INSTANCE = InnerStaticSingleton()
    }
}